packages/shell-api/src/abstract-cursor.ts

```diff
@@ -24,12 +24,37 @@ export abstract class AbstractCursor extends ShellApiClass {
   abstract _cursor: ServiceProviderAggregationCursor | ServiceProviderCursor;
   _currentIterationResult: CursorIterationResult | null = null;
   _batchSize: number | null = null;
+  _mapError: Error | null = null;
 
   constructor(mongo: Mongo) {
     super();
     this._mongo = mongo;
   }
 
+  // Wrap a function with checks before and after that verify whether a .map()
+  // callback has resulted in an exception. Such an error would otherwise result
+  // in an uncaught exception, bringing the whole process down.
+  // The downside to this is that errors will not actually be visible until
+  // the caller tries to interact with this cursor in a way that triggers
+  // these checks. Since that is also the behavior for errors coming from the
+  // database server, it makes sense to match that.
+  // Ideally, this kind of code could be lifted into the driver (NODE-3231 and
+  // NODE-3232 are the tickets for that).
+  async _withCheckMapError<Ret>(fn: () => Ret): Promise<Ret> {
+    if (this._mapError) {
```

Comment:
Nice trick! is also pretty clear, i thought it would have been more hacky

Reply:
Yeah … would still be nice if the driver had something here :)